### PR TITLE
Flexible access restrictions

### PR DESCRIPTION
**Pull request by atanasormanow**
## Changes
The following format for the access field will be supported for both metrics and signals:
```json
"access": {
  "historical": "free",
  "realtime": "restricted"
}
```

## Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have tried to find clearer solution before commenting hard-to-understand parts of code
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: lib/sanbase/clickhouse/metric/metric_files/labeled_balance_metrics.json [hunk 1]
```diff
@@ -38,7 +38,10 @@
     "name": "dex_trader_balance",
     "metric": "dex_trader_balance",
     "version": "2019-01-01",
-    "access": "restricted",
```
Comment (atanasormanow) [inline lib/sanbase/clickhouse/metric/metric_files/labeled_balance_metrics.json:41]: TODO: Revert this change

File: lib/sanbase_web/graphql/middlewares/access_control.ex [hunk 2]
```diff
@@ -214,11 +214,27 @@ defmodule SanbaseWeb.Graphql.Middlewares.AccessControl do
...
+          %{allow_historical_data: AccessChecker.is_historical_data_allowed(query)},
+          historical_data_in_days
+        ),
+        restrict_to(
+          to,
+          %{allow_realtime_data: AccessChecker.is_realtime_data_allowed(query)},
+          realtime_data_cut_off_in_days
+        )
+      )
+    else
```
Comment (IvanIvanoff) [inline lib/sanbase_web/graphql/middlewares/access_control.ex:231]: Something here is not right - you can enter this block only if `query` is the one of the atoms `:metric` or `:signal`, but then you would call `AccessChecker.is_historical_data_allowed(:metric)}` which does not yield meaningful results as the original metric is lost

File: lib/sanbase/billing/plan/access_checker.ex [hunk 3]
```diff
@@ -148,27 +168,43 @@ defmodule Sanbase.Billing.Plan.AccessChecker do
...
+    else
+      Map.get(@custom_access_queries_stats, query_or_argument)
+      |> get_in([:plan_access, plan, :historical_data_in_days])
+    end
   end
 
   for {product_id, module} <- @product_to_access_module do
     def historical_data_in_days(plan, unquote(product_id), query_or_argument) do
-      unquote(module).historical_data_in_days(plan, query_or_argument)
+      if is_historical_data_allowed?(query_or_argument) do
```
Comment (atanasormanow) [inline lib/sanbase/billing/plan/access_checker.ex:181]: @IvanIvanoff this is where the argument may end up different than `:signal` and `:metric`

File: lib/sanbase/billing/plan/access_checker.ex [hunk 4]
```diff
@@ -139,6 +139,26 @@ defmodule Sanbase.Billing.Plan.AccessChecker do
...
 
+  def is_historical_data_allowed?({:metric, metric}) do
+    Sanbase.Metric.is_historical_data_allowed?(metric)
+  end
+
+  def is_historical_data_allowed?({:signal, signal}) do
+    Sanbase.Signal.is_historical_data_allowed?(signal)
+  end
+
+  def is_historical_data_allowed?(_), do: false
```
[the remark this answers is not in the document]
Comment (atanasormanow) [reply lib/sanbase/billing/plan/access_checker.ex:150]: https://github.com/santiment/sanbase2/pull/2761/files#r660544746

File: lib/sanbase_web/graphql/middlewares/access_control.ex [hunk 5]
```diff
@@ -204,21 +204,43 @@ defmodule SanbaseWeb.Graphql.Middlewares.AccessControl do
...
     realtime_data_cut_off_in_days =
-      AccessChecker.realtime_data_cut_off_in_days(plan, product_id, query)
+      AccessChecker.realtime_data_cut_off_in_days(plan, product_id, query_or_argument)
 
-    resolution
-    |> update_resolution_from_to(
-      restrict_from(from, middleware_args, historical_data_in_days),
-      restrict_to(to, middleware_args, realtime_data_cut_off_in_days)
-    )
+    case elem(query_or_argument, 0) do
```
Comment (IvanIvanoff) [inline lib/sanbase_web/graphql/middlewares/access_control.ex:219]: You can avoid calling `elem` and use pattern matching
```elixir
case query_or_argument do
  {:query, _} ->
  
  ...
  ```

File: lib/sanbase/clickhouse/metric/file_handler.ex [hunk 6]
```diff
@@ -75,6 +75,21 @@ defmodule Sanbase.Clickhouse.MetricAdapter.FileHandler do
         end
       )
     end
+
+    def access_level_to_atom(access) when is_binary(access), do: String.to_atom(access)
```
Comment (IvanIvanoff) [inline lib/sanbase/clickhouse/metric/file_handler.ex:79]: `String.to_existing_atom`

File: lib/sanbase/clickhouse/metric/file_handler.ex [hunk 7]
```diff
@@ -75,6 +75,21 @@ defmodule Sanbase.Clickhouse.MetricAdapter.FileHandler do
...
+
+    def access_level_to_atom(access) when is_map(access) do
+      Enum.into(access, %{}, fn {k, v} -> {k, String.to_atom(v)} end)
```
Comment (IvanIvanoff) [inline lib/sanbase/clickhouse/metric/file_handler.ex:82]: `String.to_existing_atom`

File: lib/sanbase/billing/plan/access_checker.ex [hunk 8]
```diff
@@ -148,27 +168,43 @@ defmodule Sanbase.Billing.Plan.AccessChecker do
...
   def historical_data_in_days(plan, _product_id, query_or_argument)
       when query_or_argument in @custom_access_queries do
-    Map.get(@custom_access_queries_stats, query_or_argument)
-    |> get_in([:plan_access, plan, :historical_data_in_days])
+    if is_historical_data_allowed?(query_or_argument) do
+      nil
+    else
+      Map.get(@custom_access_queries_stats, query_or_argument)
+      |> get_in([:plan_access, plan, :historical_data_in_days])
+    end
```
Comment (IvanIvanoff) [inline lib/sanbase/billing/plan/access_checker.ex:176]:
```suggestion
    if not is_historical_data_allowed?(query_or_argument) do
      Map.get(@custom_access_queries_stats, query_or_argument)
      |> get_in([:plan_access, plan, :historical_data_in_days])
    end
```

File: lib/sanbase/billing/plan/access_checker.ex [hunk 9]
```diff
@@ -148,27 +170,35 @@ defmodule Sanbase.Billing.Plan.AccessChecker do
           non_neg_integer() | nil
   def historical_data_in_days(plan, _product_id, query_or_argument)
       when query_or_argument in @custom_access_queries do
-    Map.get(@custom_access_queries_stats, query_or_argument)
-    |> get_in([:plan_access, plan, :historical_data_in_days])
+    if not is_historical_data_allowed?(query_or_argument) do
```
Comment (IvanIvanoff) [inline lib/sanbase/billing/plan/access_checker.ex:173]: You can add a @doc to this function and the realtime one and explain there that if the result is `nil` then no restrictions are applied.

File: lib/sanbase/signal/file_handler.ex [hunk 10]
```diff
@@ -76,6 +76,21 @@ defmodule Sanbase.Signal.FileHandler do
         end
       )
     end
+
+    def access_level_to_atom(access) when is_binary(access), do: String.to_existing_atom(access)
+
+    def access_level_to_atom(access) when is_map(access) do
+      Enum.into(access, %{}, fn {k, v} -> {k, String.to_existing_atom(v)} end)
+    end
```
Comment (IvanIvanoff) [inline lib/sanbase/signal/file_handler.ex:84]: This name is a little bit misleading as it converts to atom or to a map. It's more like `atomize_access_level_value`

**Review by IvanIvanoff — APPROVED** [review]

**Review by tspenov — APPROVED** [review]
